core/trie.go

```
@@ -0,0 +1,351 @@
+		}
+
+		var pathP **StoragePath
+		if head == 'l' {
+			pathP = &(n.left)
+		} else if head == 'r' {
+			pathP = &(n.right)
+		} else {
+			return errors.New("Malformed TrieNode bytedata")
+		}
+
+		*pathP = new(StoragePath)
+		_, err = (*pathP).ReadFrom(stream)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// State [Trie]
+type Trie struct {
+	root    *StoragePath
+	storage TrieStorage
+}
+
+// Converts a [TrieKey] to a [StoragePath] that, when followed on a [Trie] ,leads to the corresponding [TrieNode]
+func PathFromKey(k *TrieKey) *StoragePath {
+	regularK := k.ToRegular()
+	return bitset.FromWithLength(felt.Bits-1, regularK[:])
+}
+
+// Finds the set of common MSB bits in two [StoragePath] objects
+func FindCommonPath(longerPath, shorterPath *StoragePath) (*StoragePath, bool) {
+	divergentBit := uint(0)
+
+	for divergentBit <= shorterPath.Len() &&
+		longerPath.Test(longerPath.Len()-divergentBit) == shorterPath.Test(shorterPath.Len()-divergentBit) {
+		divergentBit++
+	}
+
+	commonPath := shorterPath.Clone()
+	for i := uint(0); i < shorterPath.Len()-divergentBit+1; i++ {
+		commonPath.DeleteAt(0)
+	}
+	return commonPath, divergentBit == shorterPath.Len()+1
+}
+
+// calculates node paths according to the [docs]
+//
+// [docs]: https://docs.starknet.io/documentation/develop/State/starknet-state/
+func GetSpecPath(path, parentPath *StoragePath) *StoragePath {
+	specPath := path.Clone()
+	// drop parent path, and one more MSB since left/right relation already encodes that information
+	if parentPath != nil {
+		specPath.Shrink(specPath.Len() - parentPath.Len() - 1)
+		specPath.DeleteAt(specPath.Len() - 1)
+	}
+	return specPath
+}
+
+type step = struct {
+	path *StoragePath
+	node *StorageValue
+}
+
+// Enumerates the set of [TrieNode] objects that is on a given [StoragePath]
+func (t *Trie) stepsToRoot(path *StoragePath) (steps []step) {
+	cur := t.root
+	for cur != nil {
+		node, err := t.storage.Get(cur)
+		if err != nil {
+			return
```

Should we return the error here?